### PR TITLE
Torch JAX

### PR DESCRIPTION
**Pull request by qxcv**
This PR replaces JAX with Torch in the MCE IRL code. It also removes all mention of JAX from the rest of the repo (including dependencies).

### CODE REVIEW

**Comment by AdamGleave** [comment]
This is easily the best PR title I've ever seen.

**Comment by AdamGleave** [comment]
@shwang can you review this? I think it'd be great to drop the unnecessary JAX dependency.

**Comment by shwang** [comment]
This is on my queue right now after the March branch merge (talked about this with Sam in person). Though now that I'm reminded of it I can start looking through this between PRs

**Comment by AdamGleave** [comment]
I'll take a look at this later today or tomorrow.

File: src/imitation/algorithms/tabular_irl.py [hunk 1]
```diff
@@ -2,61 +2,39 @@
...
-        env (ModelBasedEnv): a tabular, known-dynamics MDP.
-        R (None or np.array): a reward matrix. Defaults to env.reward_matrix.
+        env: a tabular, known-dynamics MDP.
+        R: a reward matrix. Defaults to env.reward_matrix.
 
     Returns:
         (V, Q, \pi) corresponding to the soft values, Q-values and MCE policy.
-        V is a 2d array, indexed V[t,s]. Q is a 3d array, indexed Q[t,s,a].
-        \pi is a 3d array, indexed \pi[t,s,a].
+            V is a 2d array, indexed V[t,s]. Q is a 3d array, indexed Q[t,s,a].
```
Comment (AdamGleave) [inline src/imitation/algorithms/tabular_irl.py:36]: Why the extra indentation?

File: src/imitation/algorithms/tabular_irl.py [hunk 2]
```diff
@@ -91,22 +69,27 @@ def mce_partition_fh(env, *, R=None):
...
+        R: reward matrix. Defaults is env.reward_matrix.
+        pi: policy to simulate. Defaults to soft-optimal policy w.r.t reward
+            matrix.
 
     Returns:
         Tuple of D (ndarray) and Dt (ndarray). D is an :math:`|S|`-dimensional
-        vector recording the expected number of times each state is visited.
-        Dt is a :math:`T*|S|`-dimensional vector recording the probability of
-        being in a given state at a given timestep.
+            vector recording the expected number of times each state is
```
Comment (AdamGleave) [inline src/imitation/algorithms/tabular_irl.py:90]: Again, why the indentation? Is this necessary for sphinx? It doesn't seem to be used in https://google.github.io/styleguide/pyguide.html#383-functions-and-methods

File: src/imitation/algorithms/tabular_irl.py [hunk 3]
```diff
@@ -129,38 +112,61 @@ def mce_occupancy_measures(env, *, R=None, pi=None):
...
+    """Squeeze a reward output tensor down to one dimension, if necessary.
+
+    Args:
+         r_output (th.Tensor): output of reward model. Can be either 1D
+            ([n_states]) or 2D ([n_states, 1]).
+
+    Returns:
+         squeezed reward of shape [n_states].
+    """
+    th.squeeze
```
Comment (AdamGleave) [inline src/imitation/algorithms/tabular_irl.py:125]:
```suggestion
```
I love that this just does nothing in Python, but is not a syntax error.

File: src/imitation/algorithms/tabular_irl.py [hunk 4]
```diff
@@ -129,38 +112,61 @@ def mce_occupancy_measures(env, *, R=None, pi=None):
...
+            MCE IRL gradient falls below this value.
+        print_interval: how often to log current loss stats (using `logging`).
+            None to disable.
 
     Returns:
-        (np.ndarray, np.ndarray): tuple of final parameters found by optimiser
-        and state occupancy measure for the final reward function. Note that
-        rmodel will also be updated with the latest parameters."""
+        state occupancy measure for the final reward function. Note
+            that `rmodel` and `optimizer` will be updated in-place during
```
Comment (AdamGleave) [inline src/imitation/algorithms/tabular_irl.py:160]: ditto

File: src/imitation/algorithms/tabular_irl.py [hunk 5]
```diff
@@ -171,294 +177,94 @@ def mce_irl(
     # number of optimisation steps taken
     t = 0
     assert demo_state_om.shape == (len(obs_mat),)
-    opt_init, opt_update, opt_get_params = optimiser_tuple
-    rew_params = rmodel.get_params()
-    opt_state = opt_init(rew_params)
 
     while linf_delta > linf_eps and grad_norm > grad_l2_eps:
+        # zero grad
```
Comment (AdamGleave) [inline src/imitation/algorithms/tabular_irl.py:182]: Is this comment needed? `optimizer.zero_grad()` is fairly self-explanatory.

File: src/imitation/algorithms/tabular_irl.py [hunk 6]
```diff
@@ -171,294 +177,94 @@ def mce_irl(
...
-        # gradient of reward function w.r.t parameters, with expectation taken
-        # over states
-        expert_grad = np.sum(demo_state_om[:, None] * out_grads, axis=0)
-        grad = pol_grad - expert_grad
+        predicted_r = squeeze_r(rmodel(torch_obs_mat))
+        assert predicted_r.shape == (obs_mat.shape[0],)
+        predicted_r_np = predicted_r.detach().cpu().numpy()
+        _, visitations = mce_occupancy_measures(env, R=predicted_r_np)
+
+        # Forward/back/step (grads are zeroed at the top).
```
Comment (AdamGleave) [inline src/imitation/algorithms/tabular_irl.py:193]: I found this a little hard to follow. Given that the "loss" is unusual, can you be a bit more explicit? e.g. write:
```python
# Forward/back/step (grads are zeroed at the top).
# weights_th(s) = \pi(s) - D(s)
weights_th = th.as_tensor(
    visitations - demo_state_om, dtype=dtype, device=device
)
# The "loss" is then:
#   E_\pi[r_\theta(S)] - E_D[r_\theta(S)]
loss = th.dot(weights_th, predicted_r)
# This gives the required gradient:
#   E_\pi[\nabla r_\theta(S)] - E_D[\nabla r_\theta(S)].
loss.backward()
```

File: src/imitation/algorithms/tabular_irl.py [hunk 7]
```diff
@@ -171,294 +177,94 @@ def mce_irl(
...
-        Args:
-            inputs (np.ndarray): 2D matrix of observations, like .out().
 
-        Returns:
-            (np.ndarray, np.ndarray), where first array is equivalent to return
-            value of .out() and second array is equivalent to return value of
-            .grads().
-        """
-        return self.out(inputs), self.grads(inputs)
+class MLPRewardModel(nn.Module):
```
Comment (AdamGleave) [inline src/imitation/algorithms/tabular_irl.py:242]: This feels like it should be defined elsewhere and shared between algos, but it's not compatible with vanilla RewardNet, so I think we can keep it as is for this PR and then try to clean up in https://github.com/HumanCompatibleAI/imitation/pull/311

File: src/imitation/util/util.py [hunk 8]
```diff
@@ -180,3 +181,23 @@ def endless_iter(iterable: Iterable[T]) -> Iterator[T]:
...
+    tensor_iter: Iterable[th.Tensor], ord: Union[int, float] = 2  # noqa: A002
+) -> th.Tensor:
+    """Compute the norm of a big vector that is produced one tensor chunk at a time.
+
+    Args:
+        tensor_iter: an iterable that yields tensors.
+        ord: order of the p-norm (can be any int or float except 0 and NaN).
+
+    Returns:
+        norm of the concatenated tensors."""
```
Comment (AdamGleave) [inline src/imitation/util/util.py:196]:
```suggestion
        Norm of the concatenated tensors."""
```

File: src/imitation/util/util.py [hunk 9]
```diff
@@ -180,3 +181,23 @@ def endless_iter(iterable: Iterable[T]) -> Iterator[T]:
...
+    if ord == 0:
+        raise ValueError("This function cannot compute p-norms for p=0.")
+    norms = []
+    for tensor in tensor_iter:
+        norms.append(th.norm(tensor.flatten(), p=ord))
+    norm_tensor = th.as_tensor(norms)
```
Comment (AdamGleave) [inline src/imitation/util/util.py:202]:
```suggestion
    # norm_tensor[i] = (\sum_j x_{i,j}^{ord})^{1/ord}, where j ranges over tensor_iter[i]
    norm_tensor = th.as_tensor(norms)
```

File: src/imitation/util/util.py [hunk 10]
```diff
@@ -180,3 +181,23 @@ def endless_iter(iterable: Iterable[T]) -> Iterator[T]:
...
+    return th.norm(norm_tensor, p=ord)
```
Comment (AdamGleave) [inline src/imitation/util/util.py:203]:
```suggestion
    # th.norm(norm_tensor) = (\sum_i (norm_tensor[j])^{ord})^{1/ord} 
    # = (\sum_i \sum_j x_{i,j}^ord)^{1/ord}
    # = th.norm(concatenated tensors)
    return th.norm(norm_tensor, p=ord)
```

**Review by AdamGleave — APPROVED** [review]
LGTM. I've added some comments/suggestions largely for clarity. Happy to take another look but I don't think I need to review again.

The only big thing I'd like to see changed is sharing reward models between algorithms, but I think that's better done in another PR, possibly https://github.com/HumanCompatibleAI/imitation/pull/311
